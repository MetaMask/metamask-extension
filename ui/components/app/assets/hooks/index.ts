export { default as useAssetListTokenDetection } from './useAssetListTokenDetection';
export { default as useNativeTokenBalance } from './useNativeTokenBalance';
export { default as useNetworkFilter } from './useNetworkFilter';
export { default as usePrimaryCurrencyProperties } from './usePrimaryCurrencyProperties';
export { default as useTokenDisplayInfo } from './useTokenDisplayInfo';
