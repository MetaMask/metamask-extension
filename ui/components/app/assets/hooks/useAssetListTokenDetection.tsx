import { useContext, useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { NetworkConfiguration } from '@metamask/network-controller';
import { Token } from '../types';
import {
  getAllDetectedTokensForSelectedAddress,
  getDetectedTokensInCurrentNetwork,
  getIsTokenNetworkFilterEqualCurrentNetwork,
  getSelectedAddress,
  getUseTokenDetection,
} from '../../../../selectors';
import { importAllDetectedTokens } from '../util/importAllDetectedTokens';
import {
  getCurrentChainId,
  getNetworkConfigurationsByChainId,
  getSelectedNetworkClientId,
} from '../../../../../shared/modules/selectors/networks';
import { MetaMetricsContext } from '../../../../contexts/metametrics';
import {
  MetaMetricsEventCategory,
  MetaMetricsEventName,
  MetaMetricsTokenEventSource,
} from '../../../../../shared/constants/metametrics';
import {
  AssetType,
  TokenStandard,
} from '../../../../../shared/constants/transaction';
import { addImportedTokens } from '../../../../store/actions';

const useAssetListTokenDetection = () => {
  const trackEvent = useContext(MetaMetricsContext);
  const dispatch = useDispatch();
  const detectedTokensMultichain: {
    [key: string]: Token[];
  } = useSelector(getAllDetectedTokensForSelectedAddress);
  const networkClientId = useSelector(getSelectedNetworkClientId);
  const selectedAddress = useSelector(getSelectedAddress);
  const useTokenDetection = useSelector(getUseTokenDetection);
  const currentChainId = useSelector(getCurrentChainId);
  const isOnCurrentNetwork = useSelector(
    getIsTokenNetworkFilterEqualCurrentNetwork,
  );
  const detectedTokens = useSelector(getDetectedTokensInCurrentNetwork) || [];
  const allNetworks: Record<`0x${string}`, NetworkConfiguration> = useSelector(
    getNetworkConfigurationsByChainId,
  );

  const multichainDetectedTokensLength = Object.values(
    detectedTokensMultichain || {},
  ).reduce((acc, tokens) => acc + tokens.length, 0);

  const [showDetectedTokens, setShowDetectedTokens] = useState(false);

  const handleAddImportedTokens = async (
    tokens: Token[],
    networkClientIdProp: string,
  ) => {
    await dispatch(addImportedTokens(tokens as Token[], networkClientIdProp));
  };

  const trackTokenAddedEvent = (importedToken: Token, chainId: string) => {
    trackEvent({
      event: MetaMetricsEventName.TokenAdded,
      category: MetaMetricsEventCategory.Wallet,
      sensitiveProperties: {
        token_symbol: importedToken.symbol,
        token_contract_address: importedToken.address,
        token_decimal_precision: importedToken.decimals,
        source: MetaMetricsTokenEventSource.Detected,
        token_standard: TokenStandard.ERC20,
        asset_type: AssetType.token,
        token_added_type: 'detected',
        chain_id: chainId,
      },
    });
  };

  // Add detected tokens to sate
  useEffect(() => {
    if (!useTokenDetection) {
      return;
    }

    importAllDetectedTokens(
      isOnCurrentNetwork,
      detectedTokensMultichain,
      allNetworks,
      networkClientId,
      currentChainId,
      detectedTokens,
      handleAddImportedTokens,
      trackTokenAddedEvent,
    );
  }, [
    isOnCurrentNetwork,
    selectedAddress,
    networkClientId,
    detectedTokens.length,
    multichainDetectedTokensLength,
  ]);

  return { showDetectedTokens, setShowDetectedTokens };
};

export default useAssetListTokenDetection;
