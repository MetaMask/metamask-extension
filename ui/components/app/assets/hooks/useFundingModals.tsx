import { useState } from 'react';

const useFundingModals = () => {
  const [showFundingMethodModal, setShowFundingMethodModal] = useState(false);
  const [showReceiveModal, setShowReceiveModal] = useState(false);

  const onClickReceive = () => {
    setShowFundingMethodModal(false);
    setShowReceiveModal(true);
  };
};

export default useFundingModals;
