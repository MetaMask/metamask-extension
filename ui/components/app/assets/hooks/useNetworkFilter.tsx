import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { setTokenNetworkFilter } from '../../../../store/actions';
import {
  getNetworkConfigurationIdByChainId,
  getTokenNetworkFilter,
} from '../../../../selectors';

const useNetworkFilter = () => {
  const dispatch = useDispatch();

  const allNetworks = useSelector(getNetworkConfigurationIdByChainId);
  const networkFilter = useSelector(getTokenNetworkFilter);

  useEffect(() => {
    if (process.env.PORTFOLIO_VIEW) {
      const allNetworkFilters = Object.fromEntries(
        Object.keys(allNetworks).map((chainId) => [chainId, true]),
      );

      if (Object.keys(networkFilter).length > 1) {
        dispatch(setTokenNetworkFilter(allNetworkFilters));
      }
    }
  }, [Object.keys(allNetworks).length, networkFilter, dispatch]);

  return { networkFilter };
};

export default useNetworkFilter;
