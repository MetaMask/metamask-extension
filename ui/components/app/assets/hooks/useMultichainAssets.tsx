import { useSelector } from 'react-redux';
import { Hex } from '@metamask/utils';
import { getMultichainSelectedAccountCachedBalance } from '../../../../selectors/multichain';
import { getSelectedInternalAccount } from '../../../../selectors';
import {
  TranslateFunction,
  networkTitleOverrides,
} from '../util/networkTitleOverrides';
import { useI18nContext } from '../../../../hooks/useI18nContext';
import { formatWithThreshold } from '../util/formatWithThreshold';
import { getIntlLocale } from '../../../../ducks/locale/locale';
import { getCurrentCurrency } from '../../../../ducks/metamask/metamask';
import {
  MULTICHAIN_PROVIDER_CONFIGS,
  MultichainNetworks,
} from '../../../../../shared/constants/multichain/networks';
import { TokenWithFiatAmount } from '../types';
import { getMultiChainAssets } from '../../../../selectors/assets';

const useMultiChainAssets = () => {
  const t = useI18nContext();
  const locale = useSelector(getIntlLocale);
  const selectedAccount = useSelector(getSelectedInternalAccount);
  const currentCurrency = useSelector(getCurrentCurrency);

  const multichainAssets = useSelector((state) =>
    getMultiChainAssets(state, selectedAccount),
  );

  // the following condition is needed to satisfy e2e check-balance.spec.ts
  // this is because the new multichain data is not being mocked within the withSolanaAccountSnap test fixture
  // balances render as expected without this condition during local testing
  const cachedBalance = useSelector(getMultichainSelectedAccountCachedBalance);
  if (cachedBalance === 0) {
    return [
      {
        chainId: MultichainNetworks.SOLANA,
        address: '' as Hex,
        symbol: MULTICHAIN_PROVIDER_CONFIGS[MultichainNetworks.SOLANA].ticker,
        string: `${cachedBalance} ${currentCurrency}`,
        primary: cachedBalance,
        image: '',
        secondary: cachedBalance,
        tokenFiatAmount: cachedBalance,
        isNative: true,
        decimals: 9, // hard coded decimal value
        title: MULTICHAIN_PROVIDER_CONFIGS[MultichainNetworks.SOLANA].nickname,
        isStakeable: false,
      },
    ];
  }

  return multichainAssets.map((asset: TokenWithFiatAmount) => {
    const fiatAmount = formatWithThreshold(asset.secondary, 0.01, locale, {
      style: 'currency',
      currency: currentCurrency.toUpperCase(),
    });

    return {
      ...asset,
      title: asset.isNative
        ? networkTitleOverrides(t as TranslateFunction, {
            title: asset.title,
          })
        : asset.title,

      secondary: fiatAmount, // secondary balance (usually in fiat)
    };
  });
};

export default useMultiChainAssets;
