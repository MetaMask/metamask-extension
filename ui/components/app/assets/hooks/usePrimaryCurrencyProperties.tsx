import { useSelector } from 'react-redux';
import { useUserPreferencedCurrency } from '../../../../hooks/useUserPreferencedCurrency';
import { useCurrencyDisplay } from '../../../../hooks/useCurrencyDisplay';
import { getMultichainSelectedAccountCachedBalance } from '../../../../selectors/multichain';

const usePrimaryCurrencyProperties = () => {
  const balance = useSelector(getMultichainSelectedAccountCachedBalance);

  const {
    currency: primaryCurrency,
    numberOfDecimals: primaryNumberOfDecimals,
  } = useUserPreferencedCurrency('PRIMARY', {
    ethNumberOfDecimals: 4,
    shouldCheckShowNativeToken: true,
  });

  const [, primaryCurrencyProperties] = useCurrencyDisplay(balance, {
    numberOfDecimals: primaryNumberOfDecimals,
    currency: primaryCurrency,
  });

  return {
    primaryCurrencyProperties,
  };
};

export default usePrimaryCurrencyProperties;
