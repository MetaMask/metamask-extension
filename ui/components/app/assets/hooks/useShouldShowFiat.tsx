import { useSelector } from 'react-redux';
import { useMultichainSelector } from '../../../../hooks/useMultichainSelector';
import {
  getIsTestnet,
  getSelectedAccount,
  getShowFiatInTestnets,
} from '../../../../selectors';
import { getMultichainShouldShowFiat } from '../../../../selectors/multichain';

const useShouldShowFiat = () => {
  const isTestnet = useSelector(getIsTestnet);
  const selectedAccount = useSelector(getSelectedAccount);
  const shouldShowFiat = useMultichainSelector(
    getMultichainShouldShowFiat,
    selectedAccount,
  );

  const isMainnet = !isTestnet;
  const showFiatInTestnets = useSelector(getShowFiatInTestnets);

  return shouldShowFiat && (isMainnet || (isTestnet && showFiatInTestnets));
};

export default useShouldShowFiat;
