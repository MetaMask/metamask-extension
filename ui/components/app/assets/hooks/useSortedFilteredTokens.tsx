import { useMemo } from 'react';
import { shallowEqual, useSelector } from 'react-redux';
import { sortAssets } from '../util/sort';
import { filterAssets } from '../util/filter';
import {
  getCurrentNetwork,
  getNewTokensImported,
  getPreferences,
  getSelectedAccount,
  getTokenBalancesEvm,
  getTokenExchangeRates,
} from '../../../../selectors';
import { getConversionRate } from '../../../../ducks/metamask/metamask';
import { TokenWithFiatAmount } from '../types';
import useNetworkFilter from './useNetworkFilter';

const useSortedFilteredTokens = () => {
  const currentNetwork = useSelector(getCurrentNetwork);
  const { tokenSortConfig } = useSelector(getPreferences);
  const selectedAccount = useSelector(getSelectedAccount);
  const conversionRate = useSelector(getConversionRate);
  const contractExchangeRates = useSelector(
    getTokenExchangeRates,
    shallowEqual,
  );
  const newTokensImported = useSelector(getNewTokensImported);
  const evmBalances = useSelector(getTokenBalancesEvm); // TODO: Make this chain agnostic

  // network filter to determine which tokens to show in list
  const { networkFilter } = useNetworkFilter();

  return useMemo(() => {
    const filteredAssets = filterAssets(evmBalances, [
      {
        key: 'chainId',
        opts: networkFilter,
        filterCallback: 'inclusive',
      },
    ]);

    const { nativeTokens, nonNativeTokens } = filteredAssets.reduce<{
      nativeTokens: TokenWithFiatAmount[];
      nonNativeTokens: TokenWithFiatAmount[];
    }>(
      (acc, token) => {
        if (token.isNative) {
          acc.nativeTokens.push(token);
        } else {
          acc.nonNativeTokens.push(token);
        }
        return acc;
      },
      { nativeTokens: [], nonNativeTokens: [] },
    );

    const assets = [...nativeTokens, ...nonNativeTokens];
    return sortAssets(assets, tokenSortConfig);
  }, [
    tokenSortConfig,
    networkFilter,
    conversionRate,
    contractExchangeRates,
    currentNetwork,
    selectedAccount,
    newTokensImported,
  ]);
};

export default useSortedFilteredTokens;
