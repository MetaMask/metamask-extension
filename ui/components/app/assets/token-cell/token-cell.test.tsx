import React from 'react';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { fireEvent } from '@testing-library/react';
import { useSelector } from 'react-redux';
import { renderWithProvider } from '../../../../../test/lib/render-helpers';
import { useTokenFiatAmount } from '../../../../hooks/useTokenFiatAmount';
import { getCurrentCurrency } from '../../../../ducks/metamask/metamask';
import {
  getTokenList,
  getPreferences,
  getCurrencyRates,
} from '../../../../selectors';
import {
  getMultichainCurrentChainId,
  getMultichainIsEvm,
} from '../../../../selectors/multichain';

import { useIsOriginalTokenSymbol } from '../../../../hooks/useIsOriginalTokenSymbol';
import { getIntlLocale } from '../../../../ducks/locale/locale';
import TokenCell from '.';

jest.mock('react-redux', () => {
  const actual = jest.requireActual('react-redux');

  return {
    ...actual,
    useSelector: jest.fn(),
  };
});

jest.mock('../../../../hooks/useTokenFiatAmount', () => {
  return {
    useTokenFiatAmount: jest.fn(),
  };
});

jest.mock('../../../../hooks/useIsOriginalTokenSymbol', () => {
  return {
    useIsOriginalTokenSymbol: jest.fn(),
  };
});
describe('Token Cell', () => {
  const mockState = {
    metamask: {
      marketData: {
        '0x1': {
          '0xAnotherToken': { price: 0.015 },
        },
      },
      currentCurrency: 'usd',
      currencyRates: {
        ETH: {
          conversionRate: 7.0,
        },
      },
      preferences: {},
    },
  };

  (useIsOriginalTokenSymbol as jest.Mock).mockReturnValue(true);

  // two tokens with the same symbol but different addresses
  const MOCK_GET_TOKEN_LIST = {
    '0xAddress': {
      name: 'TEST-2',
      erc20: true,
      symbol: 'TEST',
      decimals: 18,
      address: '0xAddress',
      iconUrl: './images/test_1_image.svg',
      aggregators: [],
    },
    '0xAnotherToken': {
      name: 'TEST',
      erc20: true,
      symbol: 'TEST',
      decimals: 18,
      address: '0xANoTherToKen',
      iconUrl: './images/test_image.svg',
      aggregators: [],
    },
  };

  const mockStore = configureMockStore([thunk])(mockState);

  const props = {
    address: '0xAnotherToken',
    symbol: 'TEST',
    string: '5.000',
    currentCurrency: 'usd',
    image: '',
    chainId: '0x1',
    tokenFiatAmount: 5,
    onClick: jest.fn(),
  };

  const propsLargeAmount = {
    address: '0xAnotherToken',
    symbol: 'TEST',
    string: '5000000',
    currentCurrency: 'usd',
    image: '',
    chainId: '0x1',
    tokenFiatAmount: 5000000,
    onClick: jest.fn(),
  };
  const useSelectorMock = useSelector;
  (useSelectorMock as jest.Mock).mockImplementation((selector) => {
    if (selector === getPreferences) {
      return { privacyMode: false };
    }
    if (selector === getTokenList) {
      return MOCK_GET_TOKEN_LIST;
    }
    if (selector === getMultichainCurrentChainId) {
      return '0x89';
    }
    if (selector === getMultichainIsEvm) {
      return true;
    }
    if (selector === getIntlLocale) {
      return 'en-US';
    }
    if (selector === getCurrentCurrency) {
      return 'usd';
    }
    if (selector === getCurrencyRates) {
      return { POL: '' };
    }
    return undefined;
  });
  (useTokenFiatAmount as jest.Mock).mockReturnValue('5.00');

  it('should match snapshot', () => {
    const { container } = renderWithProvider(
      <TokenCell {...props} />,
      mockStore,
    );

    expect(container).toMatchSnapshot();
  });

  it('calls onClick when clicked', () => {
    const { queryByTestId } = renderWithProvider(
      <TokenCell {...props} />,
      mockStore,
    );

    const targetElem = queryByTestId('multichain-token-list-button');

    targetElem && fireEvent.click(targetElem);

    expect(props.onClick).toHaveBeenCalled();
  });

  it('should render the correct token and filter by symbol and address', () => {
    const { getByTestId, getByAltText } = renderWithProvider(
      <TokenCell {...props} />,
      mockStore,
    );

    const image = getByAltText('TEST logo');

    expect(getByTestId('multichain-token-list-item-value')).toBeInTheDocument();
    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute('src', './images/test_image.svg');
  });

  it('should render amount with the correct format', () => {
    const { getByTestId } = renderWithProvider(
      <TokenCell {...propsLargeAmount} />,
      mockStore,
    );

    const amountElement = getByTestId('multichain-token-list-item-value');

    expect(amountElement).toBeInTheDocument();
    expect(amountElement.textContent).toBe('5,000,000 TEST');
  });
});
