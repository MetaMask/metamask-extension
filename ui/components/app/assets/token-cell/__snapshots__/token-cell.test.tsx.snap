// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Token Cell should match snapshot 1`] = `
<div>
  <div
    class="mm-box multichain-token-list-item mm-box--display-flex mm-box--gap-4 mm-box--flex-direction-column"
    data-testid="multichain-token-list-item"
  >
    <a
      class="mm-box multichain-token-list-item__container-cell multichain-token-list-item__container-cell--clickable mm-box--padding-top-2 mm-box--padding-right-4 mm-box--padding-bottom-2 mm-box--padding-left-4 mm-box--display-flex mm-box--flex-direction-row"
      data-testid="multichain-token-list-button"
      href="#"
    >
      <div
        class="mm-box mm-badge-wrapper multichain-token-list-item__badge mm-box--margin-right-4 mm-box--display-inline-block"
      >
        <div
          class="mm-box mm-text mm-avatar-base mm-avatar-base--size-md mm-avatar-token mm-text--body-sm mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full"
        >
          <img
            alt="TEST logo"
            class="mm-avatar-token__token-image"
            src="./images/test_image.svg"
          />
        </div>
        <div
          class="mm-box mm-badge-wrapper__badge-container mm-badge-wrapper__badge-container--circular-top-right"
        >
          <div
            class="mm-box mm-text mm-avatar-base mm-avatar-base--size-xs mm-avatar-network multichain-token-list-item__badge__avatar-network mm-text--body-xs mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full mm-box--border-color-transparent box--border-style-solid box--border-width-1"
          >
            ?
          </div>
        </div>
      </div>
      <div
        class="mm-box multichain-token-list-item__container-cell--text-container mm-box--display-flex mm-box--flex-direction-column mm-box--width-full"
        style="flex-grow: 1; overflow: hidden;"
      >
        <div
          class="mm-box mm-box--display-flex mm-box--gap-1 mm-box--flex-direction-row mm-box--justify-content-space-between"
        >
          <div
            class="mm-box mm-box--display-inline-block mm-box--width-1/3"
          >
            <span
              class="mm-box mm-text mm-text--body-md mm-text--font-weight-medium mm-text--ellipsis mm-box--color-text-default"
            >
              TEST
            </span>
            <div
              class="mm-box mm-box--display-flex"
            >
              <p
                class="mm-box mm-text mm-text--body-sm-medium mm-text--ellipsis mm-box--color-text-default"
                data-testid="token-increase-decrease-percentage-0xAnotherToken"
              />
            </div>
          </div>
          <div
            class="mm-box mm-box--display-flex mm-box--flex-direction-column mm-box--align-items-flex-end mm-box--width-2/3"
          >
            <p
              class="mm-box mm-text mm-text--body-md mm-text--font-weight-medium mm-text--text-align-end mm-box--width-2/3 mm-box--color-text-default"
              data-testid="multichain-token-list-item-secondary-value"
            >
              5.00
            </p>
            <p
              class="mm-box mm-text mm-text--body-sm-medium mm-text--text-align-end mm-box--color-text-alternative"
              data-testid="multichain-token-list-item-value"
            >
              5
               
              TEST
            </p>
          </div>
        </div>
        <div
          class="mm-box mm-box--display-flex mm-box--gap-1 mm-box--flex-direction-row mm-box--justify-content-space-between"
        />
      </div>
    </a>
  </div>
</div>
`;
